packages/compiler/src/halo2.rs

```diff
@@ -0,0 +1,103 @@
+use std::collections::BTreeSet;
+use std::fs::File;
+use std::io::{BufWriter, Write};
+use std::path::PathBuf;
+
+use crate::{get_accepted_state, get_max_state, CompilerError, RegexAndDFA};
+
+impl RegexAndDFA {
+    fn extract_endpoints_from_defs(defs: &BTreeSet<(usize, usize)>) -> (Vec<usize>, Vec<usize>) {
```

Comment:
This function seems to return too much number of states.
For example, if the transitions in `defs` are `(0,1), (1,2), (2,3), (3,4), (4,5)` and substrings that start/end at `(1,3), (1, 4), (1,5), (2,5)` are correct (but no substring that do not start/end at any of them is correct), this function should return `([1, 1, 1, 2], [3, 4, 5, 5])`. Otherwise, an invalid substring that does not start/end at correct states might pass the verification in halo2-regex.

Reply:
@SoraSuegami I'm passing the `substr_defs` in the this function (https://github.com/shreyas-londhe/zk-regex/blob/feat/halo2-regex/packages/compiler/src/halo2.rs#L33), so it will give the correct result.

Reply:
I am still not sure that `extract_endpoints_from_defs` returns the correct endpoints for the given `substr_defs`.
For example, the `substr_defs` for the substring "a(bc|d)e" in the regex "12a(bc|d)e3" will be `[(3,4), (4,5), (4,6), (5,6)]`.
https://cyberzhg.github.io/toolbox/min_dfa?regex=MTJhKGJjfGQpZTM=
(The actual output of the compiler might be different from the above one, but the following discussion will still hold.)
Its valid start and end states (i.e., endpoints) are, respectively, states 3 and 6 because no valid substring should start/end at any of states 4 and 5.
However, the `extract_endpoints_from_defs` function will output all states.

Reply:
As the current compiler generates DFAs for each decomposed part, I think you can extract the valid start and end states for each public decomposed part in the DFA generation process.
https://github.com/shreyas-londhe/zk-regex/blob/feat/halo2-regex/packages/compiler/src/regex.rs#L398 
Specifically, the above `graph` is the DFA for each decomposed part.
https://github.com/shreyas-londhe/zk-regex/blob/feat/halo2-regex/packages/compiler/src/regex.rs#L425
When the `public_edge.0` is `max_state_index`, its `public_edge.1` denotes the first state in the target decomposed part, so I think you can take it as the start states of the endpoints.
Also, the state in `graph` that has an accept flag will be the end states of the endpoints.
@Bisht13 Please correct me if the above description about codes in `regex.rs` is wrong.

Reply:
Understood, thanks for explaining this in much detail @SoraSuegami. The logic I think will work is that we take the accepting state from the previous graph to be the starting states for the substring, and the accepting states in the current graph to be the end states for the substring. 

One thing @Bisht13 mentioned was that we might want to reveal specific state transitions and not all. But I did not see any code in the existing halo2 implementation implementation which does that. So would love to hear your thoughts on this to get more clarity.

Reply:
> we might want to reveal specific state transitions and not all. But I did not see any code in the existing halo2 implementation implementation which does that.

A developer can also specify the revealed state transitions by manually writing them in the text file given to the halo2 circuits as the substring definition, instead of using one generated by the zk-regex compiler.